src/sacn/sockets.c

```diff
@@ -585,53 +540,165 @@ void sacn_send_unicast(sacn_ip_support_t ip_supported, const uint8_t* send_buf,
 }
 
-etcpal_error_t sacn_initialize_internal_netints(SacnInternalNetintArray* internal_netints,
+etcpal_error_t sacn_initialize_receiver_netints(SacnInternalNetintArray* receiver_netints,
                                                 SacnMcastInterface* app_netints, size_t num_app_netints)
 {
-  size_t num_valid_netints = 0u;
-  etcpal_error_t result = validate_netint_config(app_netints, num_app_netints, &num_valid_netints);
+  return init_internal_netints(receiver_netints, app_netints, num_app_netints, receiver_sys_netints,
+                               num_receiver_sys_netints);
+}
 
-  SacnMcastInterface* netints_to_use = app_netints ? app_netints : sys_netints;
-  size_t num_netints_to_use = app_netints ? num_app_netints : num_sys_netints;
+etcpal_error_t sacn_initialize_source_netints(SacnInternalNetintArray* source_netints, SacnMcastInterface* app_netints,
+                                              size_t num_app_netints)
+{
+  return init_internal_netints(source_netints, app_netints, num_app_netints, source_sys_netints,
+                               num_source_sys_netints);
+}
 
-  if (result == kEtcPalErrOk)
-  {
-#if SACN_DYNAMIC_MEM
-    internal_netints->netints = calloc(num_valid_netints, sizeof(EtcPalMcastNetintId));
+etcpal_error_t source_sockets_init()
+{
+  SACN_ASSERT(num_source_sys_netints == 0);
 
-    if (!internal_netints->netints)
-      result = kEtcPalErrNoMem;
-    else
-      internal_netints->netints_capacity = num_valid_netints;
+  size_t total_sys_netints = etcpal_netint_get_num_interfaces();
+  if (total_sys_netints == 0)
+    return kEtcPalErrNoNetints;
+
+#if SACN_DYNAMIC_MEM
+  multicast_send_sockets = calloc(total_sys_netints, sizeof(etcpal_socket_t));
+  source_sys_netints = calloc(total_sys_netints, sizeof(SacnMcastInterface));
+  if (!multicast_send_sockets || !source_sys_netints)
+    return kEtcPalErrNoMem;
+#else
+  if (total_sys_netints > SACN_MAX_NETINTS)
+    total_sys_netints = SACN_MAX_NETINTS;
 #endif
-  }
 
-  if (result == kEtcPalErrOk)
+  size_t num_valid_sys_netints = 0;
+
+  const EtcPalNetintInfo* netint_list = etcpal_netint_get_interfaces();
+  for (const EtcPalNetintInfo* netint = netint_list; netint < netint_list + total_sys_netints; ++netint)
   {
-    for (size_t read_index = 0u, write_index = 0u; read_index < num_netints_to_use; ++read_index)
+    // Get the interface IP address for logging
+    char addr_str[ETCPAL_IP_STRING_BYTES];
+    addr_str[0] = '\0';
+    if (SACN_CAN_LOG(ETCPAL_LOG_INFO))
     {
-      if (netints_to_use[read_index].status == kEtcPalErrOk)
-      {
-        memcpy(&internal_netints->netints[write_index], &netints_to_use[read_index].iface, sizeof(EtcPalMcastNetintId));
-        ++write_index;
-      }
+      etcpal_ip_to_string(&netint->addr, addr_str);
     }
 
-    internal_netints->num_netints = num_valid_netints;
+    // Create a test send socket on each network interface. If it fails, we remove that interface from the respective
+    // set.
```

Comment:
@nwagner maybe not, but the inverse is more likely: If the socket can't be created now, it's unlikely that it will be created successfully in the future. This lets us avoid constantly trying to create sockets on invalid interfaces like Mac's utun, or interfaces associated with VM software, etc. every time a universe is created.

Btw, this is an adaptation of existing logic that was already in the library at 1.0.

Reply:
There definitely isn't a guarantee that a netint will have the same status, agreed there. However I'm not sure what we should do differently other than give the application a snapshot of where the netints are at currently. Maybe if one of the sockets stop working (we get an error from a socket call), we could trigger ResetNetworking internally?

Reply:
Ok, after thinking about your points I'm fine with keeping it as is.  If the sockets stop working, odds are that anything watching for network change will soon get a signal that the NIC is down anyway.